[AIX] change system dynamic library format

mustartt opened the pull request:
Historically on AIX, almost all dynamic libraries are distributed in `.a` Big Archive Format which can consists of both static and shared objects in the same archive (e.g. `libc++abi.a(libc++abi.so.1)`). During the initial porting process, the dynamic libraries are kept as `.a` to simplify the migration, but semantically having an XCOFF object under the archive extension is wrong. For crate type `cdylib` we want to be able to distribute the libraries as archives as well. 

We are migrating to archives with the following format:
```
$ ar -t lib<name>.a
lib<name>.so
```
where each archive contains a single member that is a shared XCOFF object that can be loaded.

jieyouxu commented:
Hi @mustartt, is this PR blocked on something?

daltenty commented on src/bootstrap/src/utils/helpers.rs:
```diff
@@ -52,9 +54,22 @@ pub fn exe(name: &str, target: TargetSelection) -> String {
+/// one is in xcoff format.
+fn is_aix_dylib(path: &Path) -> bool {
+    let mut buffer = [0u8; 2];
+    match File::open(path).and_then(|mut f| f.read_exact(&mut buffer)) {
```
Reading the file magic doesn't really seem correct. On AIX, archive files which contain XCOFF members with the appropriate bits set should still be dylibs.

daltenty commented:
@rustbot author

mustartt commented:
Keeping this as draft for now. Need to investigate if dylibs should be archived into `.a` and how to properly detect between static archives and archives that contains shared objects.

mustartt committed:
aix: fix archive format

fmt

fix cfg for windows

remove unused imports

address comments

update libc to 0.2.164

fmt

remove unused imports

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11885100303/job/33114285374) [(plain)](https://github.com/rust-lang/rust/commit/7af8f2ebcd634e93e171cef50cbc16ddc06654a8/checks/33114285374/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_metadata v0.0.0 (/checkout/compiler/rustc_metadata)
error[E0433]: failed to resolve: could not find `unix` in `os`
    --> compiler/rustc_metadata/src/creader.rs:1185:34
     |
1185 |         unsafe { libloading::os::unix::Library::open(Some(&new_path), flags) }.map(|lib| lib.into())
     |                                  ^^^^ could not find `unix` in `os`
note: found an item that was configured out
    --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/libloading-0.8.5/src/os/mod.rs:22:9
     |
22   | pub mod unix;
22   | pub mod unix;
     |         ^^^^
note: the item is gated here
    --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/libloading-0.8.5/src/os/mod.rs:20:1
     |
20   | #[cfg(any(unix, libloading_docs))]
help: consider importing one of these structs
     |
3    + use libloading::Library;
     |
     |
3    + use libloading::os::windows::Library;
     |
help: if you import `Library`, refer to it directly
     |
1185 -         unsafe { libloading::os::unix::Library::open(Some(&new_path), flags) }.map(|lib| lib.into())
1185 +         unsafe { Library::open(Some(&new_path), flags) }.map(|lib| lib.into())

    Checking rustc_symbol_mangling v0.0.0 (/checkout/compiler/rustc_symbol_mangling)
    Checking rustc_transmute v0.0.0 (/checkout/compiler/rustc_transmute)
    Checking rustc_incremental v0.0.0 (/checkout/compiler/rustc_incremental)
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11885288104/job/33114754229) [(plain)](https://github.com/rust-lang/rust/commit/cc8ea94f58b150edf726b3264948ec3ccda03277/checks/33114754229/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 47)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 50)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 63)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

mustartt commented:
@rustbot review

gilamn5tr reviewed (commented):
Looks good in terms of the AIX implementation.

jieyouxu commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -1142,6 +1164,31 @@ fn format_dlopen_err(e: &(dyn std::error::Error + 'static)) -> String {
+        archive_member.push(".so)");
+        let new_path = path.with_extension(archive_member);
+
+        // On AIX, we need RTLD_MEMBER to dlopen an archived shared
+        // flags = RTLD_LAZY | RTLD_LOCAL | RTLD_MEMBER
+        // these are not yet in libc https://github.com/rust-lang/libc/pull/4044
+        let flags = 0x4 | 0x80000 | 0x40000;
+        return unsafe { libloading::os::unix::Library::open(Some(&new_path), flags) }
+            .map(|lib| lib.into());
```
Question: is this still accurate? the libc PR was merged but I dunno if we bumped the libc version in r-l/r yet.

mustartt replied:
Thanks for pointing that out, it has been released a few days ago. Updating libc version

jieyouxu commented on src/bootstrap/src/utils/helpers.rs:
```diff
@@ -53,8 +54,27 @@ pub fn exe(name: &str, target: TargetSelection) -> String {
+}
+
+fn is_aix_shared_archive(path: &Path) -> bool {
+    // todo: reading the entire file as &[u8] into memory seems excessive
+    // look into either mmap it or use the &CacheReader
```
Suggestion [AIX_SHARED_ARCHIVE_HEURISTICS 2/2]: can you convert this to a FIXME and open a `C-cleanup` issue then link to it `FIXME(#nnnnnn): ...`?

jieyouxu reviewed (approved):
Thanks, this LGTM overall, left a few nits/questions.

mustartt commented on Cargo.lock:
```diff
@@ -1965,9 +1965,9 @@ checksum = "baff4b617f7df3d896f97fe922b64817f6cd9a756bb81d40f8883f2f66dcb401"
 [[package]]
 name = "libc"
-version = "0.2.161"
+version = "0.2.164"
```
`libc 0.2.163` contains the new `RTLD_MEMBER`

mustartt commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -761,14 +761,20 @@ fn link_natively(
     let (linker_path, flavor) = linker_and_flavor(sess);
     let self_contained_components = self_contained_components(sess, crate_type);
+
+    let should_archive = crate_type != CrateType::Executable && sess.target.is_like_aix;
+    let archive_member =
+        should_archive.then(|| tmpdir.join(out_filename.file_name().unwrap()).with_extension("so"));
+    let temp_filename = archive_member.as_deref().unwrap_or(out_filename);
```
Thanks added comment

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11944590366/job/33295795625) [(plain)](https://github.com/rust-lang/rust/commit/d999999dbf6b764e854805927cf62a4dafe4e774/checks/33295795625/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.9s done
#16 DONE 35.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

mustartt commented:
Rebase on newer master to avoid llvm assertion for f128. Bumped up libc to `0.2.164` that contains the new aix `RTLD_MEMBER`.

@rustbot review

jieyouxu reviewed (approved):
Thanks. My only reservation is having to load the whole shared archive in bootstrap for dylib detection, but since that only happens for `.a` I suppose it's not *too* bad, and has an associated cleanup issue https://github.com/rust-lang/rust/issues/133268, so I think that's acceptable.

jieyouxu commented:
@bors r+ rollup=never (this PR contains a libc bump)

bors commented:
:pushpin: Commit 0db9059726dbec82abdc21935f84a7f0e12ea495 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 0db9059726dbec82abdc21935f84a7f0e12ea495 -->
<!-- homu: {"type":"Approved","sha":"0db9059726dbec82abdc21935f84a7f0e12ea495","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

mustartt commented:
> Thanks. My only reservation is having to load the whole shared archive in bootstrap for dylib detection, but since that only happens for `.a` I suppose it's not _too_ bad, and has an associated cleanup issue #133268, so I think that's acceptable.

It have pretty small impact on the bootstrap time maybe an additional minute at most. But I'm wondering is the bootstrap process okay with including other dependencies such as `memmap2`? I was a bit hesitant to include it in the initial patch?

jieyouxu commented:
> But I'm wondering is the bootstrap process okay with including other dependencies such as memmap2? I was a bit hesitant to include it in the initial patch?

Ideally bootstrap needs to build fast with minimal deps, so I'm inclined to say no, not for just this one purpose.

bors commented:
:hourglass: Testing commit 0db9059726dbec82abdc21935f84a7f0e12ea495 with merge f39bacaa4ee931bea40db71b8988deff9b0c4a59...
<!-- homu: {"type":"BuildStarted","head_sha":"0db9059726dbec82abdc21935f84a7f0e12ea495","merge_sha":"f39bacaa4ee931bea40db71b8988deff9b0c4a59"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11958313566/job/33337596745) [(plain)](https://github.com/rust-lang-ci/rust/commit/f39bacaa4ee931bea40db71b8988deff9b0c4a59/checks/33337596745/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

jieyouxu commented:
@bors retry (not even trying)

bors commented:
:hourglass: Testing commit 0db9059726dbec82abdc21935f84a7f0e12ea495 with merge 5d3c6ee9b34989595d2a72b79e61ca37e949d757...
<!-- homu: {"type":"BuildStarted","head_sha":"0db9059726dbec82abdc21935f84a7f0e12ea495","merge_sha":"5d3c6ee9b34989595d2a72b79e61ca37e949d757"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11962087120/job/33355553653)
Approved by: jieyouxu
Pushing 5d3c6ee9b34989595d2a72b79e61ca37e949d757 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jieyouxu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11962087120/job/33355553653"},"merge_sha":"5d3c6ee9b34989595d2a72b79e61ca37e949d757"} -->

rust-timer commented:
Finished benchmarking commit (5d3c6ee9b34989595d2a72b79e61ca37e949d757): [comparison URL](https://perf.rust-lang.org/compare.html?start=b19329a37cedf2027517ae22c87cf201f93d776e&end=5d3c6ee9b34989595d2a72b79e61ca37e949d757&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b19329a37cedf2027517ae22c87cf201f93d776e&end=5d3c6ee9b34989595d2a72b79e61ca37e949d757&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.5% | [1.5%, 1.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b19329a37cedf2027517ae22c87cf201f93d776e&end=5d3c6ee9b34989595d2a72b79e61ca37e949d757&stat=max-rss)
<details>
<summary>Results (primary 1.6%, secondary -3.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6%  | [1.6%, 1.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.1% | [-3.4%, -2.7%] | 2     |
| All ❌✅ (primary)                 | 1.6%  | [1.6%, 1.6%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b19329a37cedf2027517ae22c87cf201f93d776e&end=5d3c6ee9b34989595d2a72b79e61ca37e949d757&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=b19329a37cedf2027517ae22c87cf201f93d776e&end=5d3c6ee9b34989595d2a72b79e61ca37e949d757&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 797.142s -> 796.222s (-0.12%)
**Artifact size**: 336.05 MiB -> 336.06 MiB (0.00%)
